datafusion/functions/src/core/nvl2.rs

```diff
@@ -0,0 +1,321 @@
+static SUPPORTED_NVL2_TYPES: &[DataType] = &[
+    DataType::Boolean,
+    DataType::UInt8,
+    DataType::UInt16,
+    DataType::UInt32,
+    DataType::UInt64,
+    DataType::Int8,
+    DataType::Int16,
+    DataType::Int32,
+    DataType::Int64,
+    DataType::Float32,
+    DataType::Float64,
+    DataType::Utf8,
+    DataType::LargeUtf8,
+];
+
+impl NVL2Func {
+    pub fn new() -> Self {
+        Self {
+            signature:
+            Signature::uniform(3, SUPPORTED_NVL2_TYPES.to_vec(),
+                Volatility::Immutable,
+            ),
+        }
+    }
+}
+
+impl ScalarUDFImpl for NVL2Func {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "nvl2"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
+        if arg_types.len() != 3 {
+            return Err(plan_datafusion_err!(
+                "{}",
+                utils::generate_signature_error_msg(
+                    self.name(),
+                    self.signature().clone(),
+                    arg_types,
+                )
+            ));
+        }
+        Ok(arg_types[1].clone())
+    }
+
+    fn invoke(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
+        nvl2_func(args)
+    }
+}
+
+fn nvl2_func(args: &[ColumnarValue]) -> Result<ColumnarValue> {
+    if args.len() != 3 {
+        return internal_err!(
+            "{:?} args were supplied but NVL2 takes exactly three args",
+            args.len()
+        );
+    }
+    let mut len = 1;
+    let mut is_array = false;
+    for arg in args {
+        if let ColumnarValue::Array(array) = arg {
+            len = array.len();
+            is_array = true;
+            break;
+        }
+    }
+    if is_array {
+        let args = args.iter().map(|arg| match arg {
+            ColumnarValue::Scalar(scalar) => {
+                scalar.to_array_of_size(len)
+            }
+            ColumnarValue::Array(array) => {
+                Ok(array.clone())
+            }
+        }).collect::<Result<Vec<_>>>()?;
+        let to_apply = is_not_null(&args[0])?;
+        let value = zip(&to_apply, &args[1], &args[2])?;
+        Ok(ColumnarValue::Array(value))
+    } else {
+        let mut current_value = &args[1];
+        match &args[0] {
+            ColumnarValue::Array(_) => {
+                internal_err!("except Scalar value, but got Array")
+            }
+            ColumnarValue::Scalar(scalar) => {
+                if scalar.is_null() {
+                    current_value = &args[2];
+                }
+                Ok(current_value.clone())
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use arrow::array::*;
+
+    use super::*;
+    use datafusion_common::{Result, ScalarValue};
+
+    #[test]
+    fn nvl2_int32() -> Result<()> {
```

Comment:
Do these tests add any additional coverage over what is in the `nvl2.slt` file? I feel like they are the same test logic encoded twice (as in it seems like any bug would cause both the slt and the rust tests to fail)

If they are the same, I think we should remove the rust based tests and stay with slt

Reply:
yes, This is indeed two similar tests, I will remove rust based tests